libsolidity/formal/CHC.cpp

```diff
@@ -1164,49 +1223,235 @@ void CHC::checkAssertTarget(ASTNode const* _scope, CHCVerificationTarget const&
+				}
+				auto const& outParams = calledFun->returnParameters();
+				initLocals += inParams.size();
+				/// In this loop we are interested in outputVars.
+				for (unsigned i = initLocals; i < initLocals + outParams.size(); ++i)
+				{
+					auto param = outParams.at(i - initLocals);
+					if (param->type()->isValueType())
+						localState += param->name() + " = " + summaryNode.second.at(i) + "\n";
+				}
+			}
+		}
+		else
+			/// We report the state after every tx in the trace except for the last, which is reported
+			/// first in the code above.
+			path.emplace_back("State: " + generatePostStateCounterexample(stateVars, calledFun, summaryNode.second));
+
+		string txCex = calledContract ? "constructor()" : generatePreTxCounterexample(stateVars, *calledFun, summaryNode.second);
+		path.emplace_back(txCex);
+
+		/// Recurse on the next interface node which represents the previous transaction
+		/// or stop.
+		if (interfaceId)
+			node = *interfaceId;
+		else
+			break;
+	}
+
+	return localState + "\nTransaction trace:\n" + boost::algorithm::join(boost::adaptors::reverse(path), "\n");
+}
+
+string CHC::generatePostStateCounterexample(vector<VariableDeclaration const*> const& _stateVars, FunctionDefinition const* _function, vector<string> const& _summaryValues)
+{
+	/// The signature of a function summary predicate is: summary(error, preStateVars, preInputVars, postInputVars, outputVars).
+	/// The signature of an implicit constructor summary predicate is: summary(error, postStateVars).
+	/// Here we are interested in postStateVars.
+	vector<string>::const_iterator stateFirst;
+	vector<string>::const_iterator stateLast;
+	if (_function)
+	{
+		stateFirst = _summaryValues.begin() + 1 + static_cast<int>(_stateVars.size()) + static_cast<int>(_function->parameters().size());
+		stateLast = stateFirst + static_cast<int>(_stateVars.size());
+	}
+	else
+	{
+		stateFirst = _summaryValues.begin() + 1;
+		stateLast = stateFirst + static_cast<int>(_stateVars.size());
+	}
+
+	vector<string> stateArgs(stateFirst, stateLast);
+	solAssert(stateArgs.size() == _stateVars.size(), "");
+
+	vector<string> stateCex;
+	for (unsigned i = 0; i < stateArgs.size(); ++i)
+	{
+		auto var = _stateVars.at(i);
+		if (var->type()->isValueType())
+			stateCex.emplace_back(var->name() + " = " + stateArgs.at(i));
+	}
+
+	return boost::algorithm::join(stateCex, ", ");
+}
+
+string CHC::generatePreTxCounterexample(vector<VariableDeclaration const*> const& _stateVars, FunctionDefinition const& _function, vector<string> const& _summaryValues)
+{
+	/// The signature of a function summary predicate is: summary(error, preStateVars, preInputVars, postInputVars, outputVars).
+	/// Here we are interested in preInputVars.
+	vector<string>::const_iterator first = _summaryValues.begin() + static_cast<int>(_stateVars.size()) + 1;
+	vector<string>::const_iterator last = first + static_cast<int>(_function.parameters().size());
```

Comment:
We said in the call these should eventually end up in the abstraction and that will probably also make the iterator juggling here go away ;-). But until then maybe add one or two assertions about the ranges you end up with being valid here?